File: 1631-path-with-minimum-effort/1631-path-with-minimum-effort.cpp
class Solution {
public:
    int minimumEffortPath(vector<vector<int>>& heights) {
        return solve(heights);
    }
private:
    // Daily
    int solve(vector<vector<int>>& heights) {
        int row = heights.size(), col = heights[0].size();
        
        // All four directions
        vector<int> dir = {0, -1, 0, 1, 0};
        
        // Stores effort for reaching a cell from (0, 0)
        vector<vector<int>> effort(row, vector<int>(col, INT_MAX));
        
        queue<pair<int, int>> queue;
        queue.push({0, 0});
        effort[0][0] = 0;
        
        while(!queue.empty()) {
            auto currCell = queue.front();
            queue.pop();
            
            // iterate for all directions
            for(int i = 0; i < 4; ++i) {
                int x = currCell.first + dir[i];
                int y = currCell.second + dir[i + 1];
                
                // if index is not feasible
                if(x < 0 || y < 0 || x >= row || y >= col)
                    continue;
                
                // if we have already less effort
                if(effort[x][y] <= effort[currCell.first][currCell.second])
                    continue;
                
                // Calculate cell difference for currCell and its adjacent cells
                int cellDiff = abs(heights[x][y] - heights[currCell.first][currCell.second]);
                
                // Maximum route effort
                int routeEffort = max(effort[currCell.first][currCell.second], cellDiff);
                
                effort[x][y] = min(effort[x][y], routeEffort);
                
                // push the adjacent cell to queue
                queue.push({x, y});
            }
        }
        return effort[row - 1][col - 1];
    }
};

File: 1631-path-with-minimum-effort/1631-path-with-minimum-effort.py
class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        row, col = len(heights), len(heights[0])
        
        # Effort matrix which stores effort for (x, y) cell from (0, 0)
        # initally all efforts are maximum
        effort = [[float('inf')] * col for _ in range(row)]
        
        # All four directions
        directions = [0, -1, 0, 1, 0]
        queue = deque()
        
        # Start with (0, 0)
        queue.append((0, 0))
        effort[0][0] = 0
        
        while queue:
            currCell = queue.popleft()
            
            # Iterate for all four directions
            for i in range(4):
                
                # indices for adjacent cells
                x = currCell[0] + directions[i]
                y = currCell[1] + directions[i + 1]
                
                # If indices are not feasible then continue
                if x < 0 or y < 0 or x >= row or y >= col:
                    continue
                
                # If effort for (x, y) is already less then continue
                if effort[x][y] <= effort[currCell[0]][currCell[1]]:
                    continue
                
                # Cell difference for current cell and its adjacent cell
                cellDiff = abs(heights[x][y] - heights[currCell[0]][currCell[1]])
                
                # maximum route effort for current cell
                routeEffort = max(effort[currCell[0]][currCell[1]], cellDiff)
                
                # Update to minimum effort
                effort[x][y] = min(effort[x][y], routeEffort)
                
                # Enqueue the adjacent cell
                queue.append((x, y))
        
        # Return the value of last cell
        return effort[row - 1][col - 1]
                
            
            